### PR TITLE
Add helperGetApiVersionFromEnv() to get apiVersion from KUBERNETES_EX…

### PR DESCRIPTION
**Pull request by everjing**
Got this error message when run 'kubectl get no' with this build locally. 
image

### CODE REVIEW

**Comment by everjing** [comment]
With this commit the build goes well with v1beta1, and I was trying to test a scenario using v1 version but it seems even the most recent aks version (1.24.6) is still using v1beta1. So I did not find a way to test v1 cases with the code.

File: pkg/token/execCredentialWriter.go [hunk 1]
```diff
@@ -20,20 +39,67 @@ type execCredentialWriter struct{}
...
+	//os.Setenv(execInfoEnv, "TEST")
+	//fmt.Println("KUBERNETES_EXEC_INFO:", os.Getenv(execInfoEnv))
+	apiVersionFromEnv, err := helperGetApiVersionFromEnv()
+	if err != nil {
+		return err
 	}
+
+	var ec interface{}
+	t := metav1.NewTime(token.Expires())
+	if apiVersionFromEnv == "client.authentication.k8s.io/v1beta1" {
```
Comment (weinong) [inline pkg/token/execCredentialWriter.go:52]: use a switch to determine the api version. the rest of the code can be outside the switch block to avoid duplication

File: pkg/token/execCredentialWriter.go [hunk 2]
```diff
@@ -20,20 +39,67 @@ type execCredentialWriter struct{}
...
 	if err := e.Encode(ec); err != nil {
 		return fmt.Errorf("could not write the ExecCredential: %s", err)
 	}
 	return nil
 }
+
+func helperGetApiVersionFromEnv() (string, error) {
+	env := os.Getenv(execInfoEnv)
+	if env == "" {
+		return "client.authentication.k8s.io/v1beta1", nil
```
Comment (weinong) [inline pkg/token/execCredentialWriter.go:86]: use const

File: pkg/token/execCredentialWriter.go [hunk 3]
```diff
@@ -20,20 +39,67 @@ type execCredentialWriter struct{}
...
+		if err := scheme.Convert(obj, &execCredential, nil); err != nil {
+			return "", fmt.Errorf("cannot convert to ExecCredential: %w", err)
+		}
+		if execCredential.TypeMeta.APIVersion == "" {
+			return "client.authentication.k8s.io/v1beta1", nil
+		}
+		if execCredential.TypeMeta.APIVersion == "client.authentication.k8s.io/v1beta1" || execCredential.TypeMeta.APIVersion == "client.authentication.k8s.io/v1" {
+			return execCredential.TypeMeta.APIVersion, nil
+		} else {
+			return "", errors.New("This api version is not supported")
```
Comment (weinong) [inline pkg/token/execCredentialWriter.go:102]: error message should indicate the provided api version

File: pkg/token/execCredentialWriter.go [hunk 4]
```diff
@@ -20,20 +39,67 @@ type execCredentialWriter struct{}
...
+	} else {
+		obj, _, err := codecs.UniversalDeserializer().Decode([]byte(env), nil, nil)
+		if err != nil {
+			return "", fmt.Errorf("decode: %w", err)
+		}
+		var execCredential clientauthentication.ExecCredential
+		if err := scheme.Convert(obj, &execCredential, nil); err != nil {
+			return "", fmt.Errorf("cannot convert to ExecCredential: %w", err)
+		}
+		if execCredential.TypeMeta.APIVersion == "" {
```
Comment (weinong) [inline pkg/token/execCredentialWriter.go:96]: switch will look cleaner

File: pkg/token/execCredentialWriter.go [hunk 5]
```diff
@@ -20,20 +28,63 @@ type execCredentialWriter struct{}
...
+	}
+
 	e := json.NewEncoder(os.Stdout)
 	if err := e.Encode(ec); err != nil {
 		return fmt.Errorf("could not write the ExecCredential: %s", err)
 	}
 	return nil
 }
+
+func helperGetApiVersionFromEnv() (string, error) {
```
Comment (weinong) [inline pkg/token/execCredentialWriter.go:70]: nit: you don't need helper prefix. how about `getAPIVersionFromExecInfoEnv`

File: pkg/token/execCredentialWriter.go [hunk 6]
```diff
@@ -20,20 +28,63 @@ type execCredentialWriter struct{}
...
+	env := os.Getenv(execInfoEnv)
+	if env == "" {
+		return apiV1beta1, nil
+	}
+	var execCredential clientauthentication.ExecCredential
+	error := json.Unmarshal([]byte(env), &execCredential)
+	if error != nil {
+		return "", fmt.Errorf("cannot convert to ExecCredential: %w", error)
```
Comment (weinong) [inline pkg/token/execCredentialWriter.go:78]:
```suggestion
		return "", fmt.Errorf("cannot unmarshall %q to ExecCredential: %w", env, error)
```

File: pkg/token/execCredentialWriter.go (hunk 5 above)
Comment (everjing) [reply pkg/token/execCredentialWriter.go]: Changed in new commit d64de9f. Please check it out. Thanks!

File: pkg/token/execCredentialWriter.go (hunk 6 above)
Comment (everjing) [reply pkg/token/execCredentialWriter.go]: Changed per your suggestion in new commit d64de9f. Please check it out. Thanks!

File: pkg/token/execCredentialWriter.go [hunk 7]
```diff
@@ -20,20 +28,63 @@ type execCredentialWriter struct{}
...
+	var execCredential clientauthentication.ExecCredential
+	error := json.Unmarshal([]byte(env), &execCredential)
+	if error != nil {
+		return "", fmt.Errorf("cannot unmarshall %q to ExecCredential: %w", env, error)
+	}
+	switch execCredential.TypeMeta.APIVersion {
+	case "":
+		return apiV1beta1, nil
+	case apiV1:
+		return execCredential.TypeMeta.APIVersion, nil
```
Comment (weinong) [inline pkg/token/execCredentialWriter.go:84]: fallthrough

Comment (everjing) [reply pkg/token/execCredentialWriter.go]: Use case list in switch for the most recent commit 7b8815b. Please check it out. Thanks!

**Comment by weinong** [comment]
Generally looks good to me. but can you please add unit tests?

File: pkg/token/execCredentialPlugin_test.go [hunk 8]
```diff
@@ -12,7 +12,7 @@ import (
 	"github.com/golang/mock/gomock"
 )
 
-func TestExecCredentialPlugin(t *testing.T) {
+func TestExecCredentialWriter(t *testing.T) {
```
Comment (weinong) [inline pkg/token/execCredentialPlugin_test.go:15]: why is this change needed?

File: pkg/token/execCredentialWriter.go [hunk 9]
```diff
@@ -3,37 +3,91 @@ package token
...
+			},
+			Status: &v1.ExecCredentialStatus{
+				Token:               token.AccessToken,
+				ExpirationTimestamp: &t,
+			},
+		}
+	}
+	var ecCopy interface{} = ec
+	content, _ := json.Marshal(ecCopy)
+	//fmt.Fprintln(os.Stderr, string(content))
```
Comment (weinong) [inline pkg/token/execCredentialWriter.go:65]: you can use klog with higher verbosity level for logging.

File: pkg/token/execCredentialWriter.go [hunk 10]
```diff
@@ -3,37 +3,91 @@ package token
...
 
+const (
+	apiV1       string = "client.authentication.k8s.io/v1"
+	apiV1beta1  string = "client.authentication.k8s.io/v1beta1"
+	execInfoEnv string = "KUBERNETES_EXEC_INFO"
+)
+
 type ExecCredentialWriter interface {
-	Write(token adal.Token) error
+	Write(token adal.Token, buffer bytes.Buffer) error
```
Comment (weinong) [inline pkg/token/execCredentialWriter.go:25]:
```suggestion
	Write(token adal.Token, writer io.Writer) error
```
it should be like this so that you pass in os.Stdout from kubelogin and pass in buf.Buffer from unit test

File: pkg/token/execCredentialWriterAPIVersion_test.go [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+	"github.com/Azure/go-autorest/autorest/adal"
+	"k8s.io/client-go/pkg/apis/clientauthentication"
+)
+
+
+func TestExecCredentialWriterAPIVersion() (t *testing.T) {
+	testData1 := []struct {
+		name               string
+		execInfoEnvTest    string
+		stringBufferTest   bytes.Buffer
```
Comment (weinong) [inline pkg/token/execCredentialWriterAPIVersion_test.go:20]: you don't need it in the test data, just create a local bytes.Buffer after L59

File: pkg/token/execCredentialWriterAPIVersion_test.go [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+		expectedAPIVersion string
+	}{
+		{
+			name: "KUBERNETES_EXEC_INFO is empty",
```
Comment (weinong) [inline pkg/token/execCredentialWriterAPIVersion_test.go:24]: typical test case should describe the scenario and the expected outcome

File: pkg/token/execCredentialWriterAPIVersion_test.go (hunk 11 above)
Comment (weinong) [reply pkg/token/execCredentialWriterAPIVersion_test.go:20]: besides, https://pkg.go.dev/bytes#NewBuffer, 
> In most cases, new(Buffer) (or just declaring a Buffer variable) is sufficient to initialize a Buffer.

File: pkg/token/execCredentialWriterAPIVersion_test.go [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+
+	}
+
+	for _, data := range testData1 {
+		t.Run(data.name, func(t *testing.T) {
+			os.SetEnv("KUBERNETES_EXEC_INFO", testData.execInfoEnvTest)
+			defer os.UnsetEnv("KUBERNETES_EXEC_INFO")
+			execCredentialWriter.Write(adal.Token, data.stringBufferTest)
+			var execCredential clientauthentication.ExecCredential
+			json.Unmarshal([]byte(data.stringBufferTest.String()), &execCredential)
```
Comment (weinong) [inline pkg/token/execCredentialWriterAPIVersion_test.go:62]: .Bytes() will return []byte

File: pkg/token/execCredentialWriterAPIVersion_test.go (hunk 11 above)
Comment (everjing) [reply pkg/token/execCredentialWriterAPIVersion_test.go]: Addressed these two in the most recent commit 244a97c

File: pkg/token/execCredentialWriterAPIVersion_test.go (hunk 12 above)
Comment (everjing) [reply pkg/token/execCredentialWriterAPIVersion_test.go]: As the test logic for this is relatively easy so I put these in t.Run() func.

**Comment by weinong** [comment]
please remove the extra generated mock files...

File: pkg/token/execCredentialPlugin_test.go (hunk 8 above)
Comment (weinong) [reply pkg/token/execCredentialPlugin_test.go:16]: @everjing, this file is to test execCredentialPlugin, why is this function changed?

File: pkg/token/execCredentialWriter.go [hunk 14]
```diff
@@ -3,37 +3,92 @@ package token
 //go:generate sh -c "mockgen -destination mock_$GOPACKAGE/execCredentialWriter.go github.com/Azure/kubelogin/pkg/token ExecCredentialWriter"
 
 import (
+	//"bytes"
```
Comment (weinong) [inline pkg/token/execCredentialWriter.go:6]: please remove unnecessary comments

File: pkg/token/execCredentialWriter.go (hunk 9 above)
Comment (weinong) [reply pkg/token/execCredentialWriter.go:66]: again, remove unnecessary commented code

File: pkg/token/execCredentialWriter.go [hunk 15]
```diff
@@ -3,37 +3,92 @@ package token
...
+
+func getAPIVersionFromExecInfoEnv() (string, error) {
+	env := os.Getenv(execInfoEnv)
+	if env == "" {
+		return apiV1beta1, nil
+	}
+	var execCredential clientauthentication.ExecCredential
+	error := json.Unmarshal([]byte(env), &execCredential)
+	if error != nil {
+		return "", fmt.Errorf("cannot unmarshall %q to ExecCredential: %w", env, error)
```
Comment (weinong) [inline pkg/token/execCredentialWriter.go:84]: typo: unmarshal

**Review by weinong — CHANGES_REQUESTED** [review]
so close! please address the comments so that we can merge it

**Comment by weinong** [comment]
you also need to rebase

File: pkg/token/execCredentialPlugin_test.go (hunk 8 above)
Comment (everjing) [reply pkg/token/execCredentialPlugin_test.go:16]: You are right, this change is not needed and this func name will be reverted in next commit. Thanks!

File: pkg/token/execCredentialWriter.go (hunk 15 above)
Comment (everjing) [reply pkg/token/execCredentialWriter.go:84]: Done for next commit. Thanks!

File: pkg/token/execCredentialWriter.go (hunk 14 above)
Comment (everjing) [reply pkg/token/execCredentialWriter.go:6]: Done for next commits. Thanks!

File: pkg/token/execCredentialWriter.go (hunk 10 above)
Comment (everjing) [reply pkg/token/execCredentialWriter.go]: Done in the current commit. Thanks!

File: pkg/token/execCredentialWriter.go (hunk 4 above)
Comment (everjing) [reply pkg/token/execCredentialWriter.go]: Done. Thanks!

File: pkg/token/execCredentialWriter.go (hunk 1 above)
Comment (everjing) [reply pkg/token/execCredentialWriter.go]: Done already. Thanks!

File: pkg/token/execCredentialWriter.go (hunk 2 above)
Comment (everjing) [reply pkg/token/execCredentialWriter.go]: Done already in current commit. Thanks!

File: pkg/token/execCredentialWriter.go (hunk 3 above)
Comment (everjing) [reply pkg/token/execCredentialWriter.go]: Done already in current commit. Thanks!

**Review by weinong — APPROVED** [review]
merge it!
